File: python/day-02/src/day_02/part_two.py
from typing import Dict, List


def read_input(fp: str) -> List[str]:
    with open(fp, "r") as file:
        return [line.strip() for line in file.readlines()]


def part_two(input: List[str]) -> int:
    output = 0
    for line in input:
        output += parse_line(line)

    return output


def parse_line(line: str) -> int:
    max_red, max_green, max_blue = 0, 0, 0
    records = line.split(": ")[1]

    for record in records.split("; "):
        cubes = parse_game_record(record)
        if cubes.get("red", 0) > max_red:
            max_red = cubes.get("red")
        if cubes.get("green", 0) > max_green:
            max_green = cubes.get("green")
        if cubes.get("blue", 0) > max_blue:
            max_blue = cubes.get("blue")

    return max_red * max_green * max_blue


def parse_game_record(record: str) -> Dict[str, int]:
    cubes = record.split(", ")
    return {cube.split()[1]: int(cube.split()[0]) for cube in cubes}


def main():
    input = read_input("input.txt")
    output = part_two(input)
    print(output)


if __name__ == "__main__":
    main()

File: rust/day-02/src/bin/part_two.rs
fn main() {
    let input = include_str!("./input.txt");
    let output = part_two(&input);
    dbg!(output);
}

fn part_two(input: &str) -> u32 {
    let output = input.lines().map(|line| parse_line(line)).sum::<u32>();

    output
}

fn parse_line(line: &str) -> u32 {
    let (mut max_red, mut max_green, mut max_blue) = (0, 0, 0);

    let mut line_iterator = line.split(": ").skip(1);
    let records = line_iterator.next().unwrap();

    for record in records.split("; ") {
        parse_game_record(record, &mut max_red, &mut max_green, &mut max_blue);
    }

    max_red * max_green * max_blue
}

fn parse_game_record(record: &str, red: &mut u32, green: &mut u32, blue: &mut u32) {
    for cube in record.split(", ") {
        let cube_count = parse_int_from_str(cube);
        if cube.ends_with("red") {
            if cube_count > *red {
                *red = cube_count;
            }
        } else if cube.ends_with("green") {
            if cube_count > *green {
                *green = cube_count;
            }
        } else {
            if cube_count > *blue {
                *blue = cube_count;
            }
        }
    }
}

fn parse_int_from_str(str: &str) -> u32 {
    let int = str
        .chars()
        .filter(|char| char.is_ascii_digit())
        .collect::<String>()
        .parse::<u32>()
        .unwrap();

    int
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part_two() {
        let result = part_two(
            "Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green",
        );
        assert_eq!(result, 2286)
    }
}
